### PR TITLE
ENH: Make `np.in1d()` work for unorderable object arrays

### LINKED ISSUE #9914: BUG: np.isin and np.in1d fails on unsortable object arrays
This is a problem for object arrays that are unsortable, such as those which
* are heterogenous (py 3)
* contain `None` (py 3)
* contain objects where `cmp(a, b) == 1` (#9874)

`np.isin` uses sorting for efficiency on large arrays, but in these cases this isn't valid. For example:

```python
>>> np.isin(1, [None]*5)
array(False, dtype=bool)
>>> np.isin(1, [None]*10)
TypeError: unorderable types: NoneType() < NoneType()
```

In the latter case (#9874) the problem is even worse, as a wrong result is silently produced.

I don't think there's any option here but to drop the efficient behaviour when `dtype == object`

### PR DESCRIPTION
**Pull request by jdoepfert**
Fixes #9914 

`np.in1d()` sorts large input arrays for efficiency reasons. However, the sorting throws an error when one of the input arrays contains objects that are not orderable:

```python
>>> np.isin(1, [None]*10)
TypeError: unorderable types: NoneType() < NoneType()
>>> np.isin([None], [1]*10)
TypeError: unorderable types: int() < NoneType()
```

Fix: Do not use sorting whenever one of the input arrays may contain arbitrary objects (`dtype.hasobject == True`).

### CODE REVIEW

File: numpy/lib/arraysetops.py [hunk 1]
```diff
@@ -448,8 +448,14 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
     ar1 = np.asarray(ar1).ravel()
     ar2 = np.asarray(ar2).ravel()
 
-    # This code is significantly faster when the condition is satisfied.
-    if len(ar2) < 10 * len(ar1) ** 0.145:
+    # Check if one of the arrays is of type object
+    is_object = ar1.dtype == np.dtype('O') or ar2.dtype == np.dtype('O')
```
Comment (eric-wieser) [inline numpy/lib/arraysetops.py]: I think you actually want `ar1.dtype..hasobject or ar2.dtype.hasobject`, to catch structured dtypes containing arrays

Comment (eric-wieser) [reply numpy/lib/arraysetops.py]: Needs a test for this too

Comment (jdoepfert) [reply numpy/lib/arraysetops.py]: thx for the hint!

**Comment by eric-wieser** [comment]
Reiterating my hidden comment - this needs a test for a dtype like `np.dtype([('a', object), ('b', int)])`

**Comment by jdoepfert** [comment]
@eric-wieser Good point, I added a test for arrays of structured dtype containing objects

File: numpy/lib/tests/test_arraysetops.py [hunk 2]
```diff
@@ -208,6 +208,37 @@ def test_in1d_ravel(self):
...
+
+    def test_in1d_both_arrays_have_structured_dtype(self):
+        # Test arrays of a structured data type containing an integer field
+        # and a field of dtype `object` allowing for arbitrary Python objects
+        dt = np.dtype([('field1', int), ('field2', object)])
+        ar1 = np.array([(1, 2)], dtype=dt)
+        ar2 = np.array([(1, 2)]*10, dtype=dt)
+        expected = np.array([True])
+        result = np.in1d(ar1, ar2)
+        assert_array_equal(result, expected)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_arraysetops.py]: This test already passes for me on `master` - are we missing something here?

Comment (eric-wieser) [reply numpy/lib/tests/test_arraysetops.py]: Right, I remember - the types need to be non-comparable, so maybe change the `2`s to `None`s?

Comment (jdoepfert) [reply numpy/lib/tests/test_arraysetops.py]: Of course! The test now fails on my `master`.

**Review by eric-wieser — APPROVED** [review]
LGTM. Commits need squashing, but we can do that when we merge (which I'll do in a couple days assuming no one else reviews)

**Comment by jdoepfert** [comment]
👍 thanks for reviewing @eric-wieser

**Pull request merged**